src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +84,224 @@ public PrimitiveColumnContainer(long length = 0)
             {
                 Buffers.Add(new DataFrameBuffer<T>());
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
             DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
             if (lastBuffer.Length == lastBuffer.MaxCapacity)
             {
                 lastBuffer = new DataFrameBuffer<T>();
+                Buffers.Add(lastBuffer);
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
-            lastBuffer.Append(value);
+            lastBuffer.Append(value ?? default);
+            SetValidityBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public void AppendMany(T? value, long count)
+        {
+            if (!value.HasValue)
+            {
+                NullCount += count;
+            }
+            while (count > 0)
+            {
+                if (Buffers.Count == 0)
+                {
+                    Buffers.Add(new DataFrameBuffer<T>());
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
+                if (lastBuffer.Length == lastBuffer.MaxCapacity)
+                {
+                    lastBuffer = new DataFrameBuffer<T>();
+                    Buffers.Add(lastBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                int allocatable = (int)Math.Min(count, lastBuffer.MaxCapacity);
+                lastBuffer.EnsureCapacity(allocatable);
+                lastBuffer.Span.Slice(lastBuffer.Length, allocatable).Fill(value ?? default);
+                lastBuffer.Length += allocatable;
+                Length += allocatable;
+
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[NullBitMapBuffers.Count - 1];
+                int nullBitMapAllocatable = (int)Math.Ceiling(allocatable / 8.0);
+                lastNullBitMapBuffer.EnsureCapacity(nullBitMapAllocatable);
+                _modifyNullCountWhileIndexing = false;
+                for (long i = Length - count; i < Length; i++)
+                {
+                    SetValidityBit(i, value.HasValue ? true : false);
+                }
+                _modifyNullCountWhileIndexing = true;
+                lastNullBitMapBuffer.Length += nullBitMapAllocatable;
+                count -= allocatable;
+            }
+        }
+
+        public bool IsValid(long index) => NullCount == 0 || GetValidityBit(index);
+
+        /// <summary>
+        /// A null value has an unset bit
+        /// A NON-null value has a set bit
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        private void SetValidityBit(long index, bool value)
+        {
+            if ((uint)index > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
```

Comment:
ThrowHelper?

Reply:
Yup. Just not in this patch. Which namespace has ThrowHelper though? When I look at source.dot.net, it seems to be `internal` to System?

Reply:
ThrowHelper is more a pattern than a concrete class. The motivation for that use is best described in [this comment from corefx](https://github.com/dotnet/corefx/blob/a1940826e89c4c333bca45459b25ba009e869ff6/src/Common/src/CoreLib/System/ThrowHelper.cs#L6-L36). Note: it makes only sense in lower-level APIs, not in user-code.
To see the effect of the ThrowHelper-pattern see [this SharpLab-demo](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAWQAUg4QDdsAGwCuMAJRcenaj00M+AMwZiZCmAwA8DAAwqNWmxgAWUCAHcGAmC4CCUAObz8MIQB5eQxAnQAlbAFvGABRBDAYAAcMPggBMQFsfwgdAzlFJSUAblUbMq1iAHYGQ0VS624AXwqK+m0hBgA5CU665Qr1LQBIXX1+0wsrGxsAFQdnAAkYWSSYKAA6eccnL19/IPkdMMjomDFtpZW1zfjElLSBPb8AjA3+korNL55q2oKYA1NC1qCCaO1iOQkExSAxLk5lqt1jQhr9GJDocQUHCFrsfC9DscIlEYmI7slUulngdhNh8TTpjMALwAPgY9h2rncDGpr2CoWJZ3JD3SYjp+1eWwgAGUMFBBN4xEUgTw2owAn4GMLKU96a9Bl9hhNuBUQU0gA==) (compare method `M` with `N` in the dissambly and see that `N` is straightforward, whilst `M` has all the exception code in it).

In essence this pattern moves the exception code out of the method (and possible out of hot-paths), thus resulting in  more compact and faster code.

For your question this means, if this project doesn't have a `ThrowHelper`-class you can create one.
Note also that a distinct class isn't needed to take advantage of moving out the exception code from "worker-code", as this can be done with any method (even local methods). But with a distinct class all exception-code is in one place and can be reused.

At the JIT-level this advantage is based on the fact that the JIT recognises methods that won't return (as throw-codepath will not return), so it doesn't inline these methods and moves it out of the hot-path.